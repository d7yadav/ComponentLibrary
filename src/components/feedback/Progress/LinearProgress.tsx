import { forwardRef, memo } from 'react';

import {
  DEFAULT_PROGRESS_PROPS,
  ACCESSIBILITY_CONSTANTS,
  PROGRESS_VALUE_FORMATTERS,
} from './Progress.constants';
import { 
  StyledLinearProgress, 
  ProgressContainer,
  ProgressLabel,
  ProgressValue
} from './Progress.styles';
import type { LinearProgressProps } from './Progress.types';

// Extracted style objects to prevent re-renders
const progressWrapperBaseStyle = {
  display: 'flex' as const,
  alignItems: 'center' as const,
  gap: 8,
};

const createProgressWrapperStyle = (fullWidth: boolean) => ({
  ...progressWrapperBaseStyle,
  width: fullWidth ? '100%' : 'auto',
});

/**
 * Enhanced Linear Progress component with comprehensive styling and accessibility
 * 
 * Features:
 * - 4 variants: determinate, indeterminate, buffer, query
 * - Multiple sizes: small, medium, large
 * - Color themes: primary, secondary, error, warning, info, success
 * - Striped and animated patterns
 * - Value display with custom formatters
 * - Accessibility compliance (WCAG 2.1 AA)
 * - TypeScript support with strict typing
 * - Dark theme support via enhanced theme system
 */
export const LinearProgress = forwardRef<HTMLDivElement, LinearProgressProps>((
  {
    variant = DEFAULT_PROGRESS_PROPS.variant,
    value,
    valueBuffer,
    color = DEFAULT_PROGRESS_PROPS.color,
    size = DEFAULT_PROGRESS_PROPS.size,
    thickness,
    label,
    showValue = DEFAULT_PROGRESS_PROPS.showValue,
    valueFormat = PROGRESS_VALUE_FORMATTERS.percentage,
    animated = DEFAULT_PROGRESS_PROPS.animated,
    animationDuration,
    striped = DEFAULT_PROGRESS_PROPS.striped,
    stripedAnimated = DEFAULT_PROGRESS_PROPS.stripedAnimated,
    bgcolor,
    progressColor,
    fullWidth = DEFAULT_PROGRESS_PROPS.fullWidth,
    height,
    rounded = DEFAULT_PROGRESS_PROPS.rounded,
    borderRadius,
    elevated = DEFAULT_PROGRESS_PROPS.elevated,
    elevation,
    className,
    sx,
    'aria-label': ariaLabel,
    'aria-describedby': ariaDescribedBy,
    'aria-valuetext': ariaValueText,
    ...other
  },
  ref
) => {
  // Format the value for display
  const formatValue = (val?: number) => {
    if (val === undefined) return '';
    return valueFormat(val);
  };

  // Accessibility props
  const accessibilityProps = {
    role: ACCESSIBILITY_CONSTANTS.progressRole,
    'aria-label': ariaLabel || ACCESSIBILITY_CONSTANTS.progressLabel,
    'aria-describedby': ariaDescribedBy,
    'aria-valuemin': ACCESSIBILITY_CONSTANTS.minValue,
    'aria-valuemax': ACCESSIBILITY_CONSTANTS.maxValue,
    'aria-valuenow': value,
    'aria-valuetext': ariaValueText || (value !== undefined ? formatValue(value) : undefined),
  };

  const progressElement = (
    <StyledLinearProgress
      ref={ref}
      variant={variant}
      value={value}
      valueBuffer={valueBuffer}
      color={color}
      className={className}
      sx={sx}
      // Custom props for styled component
      customSize={size}
      customHeight={height || thickness}
      customFullWidth={fullWidth}
      customRounded={rounded}
      customBorderRadius={borderRadius}
      customElevated={elevated}
      customElevation={elevation}
      customAnimated={animated}
      customAnimationDuration={animationDuration}
      customStriped={striped}
      customStripedAnimated={stripedAnimated}
      customBgcolor={bgcolor}
      customProgressColor={progressColor}
      {...accessibilityProps}
      {...other}
    />
  );

  // If no label or value display, return just the progress bar
  if (!label && !showValue) {
    return progressElement;
  }

  // Return progress with label and/or value
  return (
    <ProgressContainer data-testid="linearprogress" customCentered={false}>
      {label && (
        <ProgressLabel customSize={size} className="progress-label-top">
          {label}
        </ProgressLabel>
      )}
      <div style={createProgressWrapperStyle(fullWidth)}>
        {progressElement}
        {showValue && value !== undefined && (
          <ProgressValue customSize={size}>
            {formatValue(value)}
          </ProgressValue>
        )}
      </div>
    </ProgressContainer>
  );
});

LinearProgress.displayName = 'LinearProgress';