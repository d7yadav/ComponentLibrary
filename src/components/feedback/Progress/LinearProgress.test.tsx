/**
 * @fileoverview LinearProgress Component Tests
 * @author AI Generated Tests - 2025-07-13
 */

import { ThemeProvider , createTheme } from '@mui/material/styles';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { axe, toHaveNoViolations } from 'jest-axe';

import { LinearProgress } from './LinearProgress';
import type { LinearProgressProps } from './LinearProgress.types';

// Add jest-axe matcher
expect.extend(toHaveNoViolations);

// Test theme
const testTheme = createTheme();

// Test wrapper component
const TestWrapper = ({ children }: { children: React.ReactNode }) => (
  <ThemeProvider theme={testTheme}>{children}</ThemeProvider>
);

// Helper function to render component with theme
const renderWithTheme = (ui: React.ReactElement, options = {}) => {
  return render(ui, { wrapper: TestWrapper, ...options });
};

describe('LinearProgress Component', () => {
  // Basic rendering tests
  describe('Rendering', () => {
    it('renders without crashing', () => {
      renderWithTheme(<LinearProgress>Test content</LinearProgress>);
      expect(screen.getByRole('progressbar')).toBeInTheDocument();
          });

    it('renders children correctly', () => {
      const testContent = 'Test LinearProgress Content';
      renderWithTheme(<LinearProgress>{testContent}</LinearProgress>);
      expect(screen.getByText(testContent)).toBeInTheDocument();
    });

    it('renders with custom test id', () => {
      const testId = 'custom-linear-progress';
      renderWithTheme(
        <LinearProgress data-testid={testId}>Test</LinearProgress>
      );
      expect(screen.getByTestId(testId)).toBeInTheDocument();
    });

    it('applies custom className', () => {
      const customClass = 'custom-class';
      renderWithTheme(
        <LinearProgress className={customClass}>Test</LinearProgress>
      );
      expect(screen.getByRole('progressbar')).toHaveClass(customClass);
    });
  });

  // Props testing
  describe('Props', () => {
    describe('variant prop', () => {
      const variants: LinearProgressProps['variant'][] = [
        'determinate',
        'indeterminate',
        'buffer',
        'query',
      ];

      variants.forEach((variant) => {
        it(`renders ${variant} variant correctly`, () => {
          const props: any = { variant, 'data-testid': 'variant-test' };
          
          // Add required props for specific variants
          if (variant === 'determinate') {
            props.value = 50;
          } else if (variant === 'buffer') {
            props.value = 50;
            props.valueBuffer = 70;
          }
          
          renderWithTheme(
            <LinearProgress {...props}>
              {variant} variant
            </LinearProgress>
          );
          const element = screen.getByTestId('variant-test');
          expect(element).toBeInTheDocument();
        });
      });
    });

    describe('size prop', () => {
      const sizes: LinearProgressProps['size'][] = [
        'small',
        'medium',
        'large',
      ];

      sizes.forEach((size) => {
        it(`renders ${size} size correctly`, () => {
          renderWithTheme(
            <LinearProgress size={size} data-testid="size-test">
              {size} size
            </LinearProgress>
          );
          const element = screen.getByTestId('size-test');
          expect(element).toBeInTheDocument();
        });
      });
    });



    it('handles disabled state correctly', () => {
      const handleClick = vi.fn();
      renderWithTheme(
        <LinearProgress disabled onClick={handleClick} data-testid="disabled-test">
          Disabled LinearProgress
        </LinearProgress>
      );
      
      const element = screen.getByTestId('disabled-test');
      expect(element).toHaveAttribute('aria-disabled', 'true');
    });
  });

  // Interaction testing
  describe('Interactions', () => {
    it('does not interfere with parent interactions', () => {
      const handleClick = vi.fn();
      renderWithTheme(
        <div onClick={handleClick} data-testid="parent">
          <LinearProgress data-testid="progress">
            Progress in clickable parent
          </LinearProgress>
        </div>
      );
      
      const parentElement = screen.getByTestId('parent');
      fireEvent.click(parentElement);
      expect(handleClick).toHaveBeenCalledTimes(1);
    });

    it('maintains aria-busy state during progress', () => {
      renderWithTheme(
        <LinearProgress variant="indeterminate" data-testid="busy-progress">
          Loading...
        </LinearProgress>
      );
      
      const element = screen.getByRole('progressbar');
      expect(element).toBeInTheDocument();
    });
  });



  // Accessibility testing
  describe('Accessibility', () => {
    it('meets WCAG accessibility guidelines', async () => {
      const { container } = renderWithTheme(
        <LinearProgress>Accessible LinearProgress</LinearProgress>
      );
      
      const results = await axe(container);
      expect(results).toHaveNoViolations();
    });

    it('supports screen readers', () => {
      const ariaLabel = 'Screen reader label';
      renderWithTheme(
        <LinearProgress aria-label={ariaLabel}>
          LinearProgress for screen readers
        </LinearProgress>
      );
      
      expect(screen.getByLabelText(ariaLabel)).toBeInTheDocument();
    });

    it('supports keyboard navigation', () => {
      renderWithTheme(<LinearProgress>Keyboard LinearProgress</LinearProgress>);
      
      const element = screen.getByRole('progressbar');
      element.focus();
      expect(element).toHaveFocus();
          });

    it('provides proper ARIA attributes', () => {
      renderWithTheme(
        <LinearProgress 
          aria-describedby="description"
          aria-controls="controlled-element"
        >
          ARIA LinearProgress
        </LinearProgress>
      );
      
      const element = screen.getByRole('progressbar');
      expect(element).toHaveAttribute('aria-describedby', 'description');
      expect(element).toHaveAttribute('aria-controls', 'controlled-element');
    });
  });

  // Theme integration testing
  describe('Theme Integration', () => {
    it('respects theme customizations', () => {
      const customTheme = createTheme({
        components: {
          MuiLinearProgress: {
            defaultProps: {
              variant: 'determinate',
            },
            styleOverrides: {
              root: {
                backgroundColor: 'rgb(255, 0, 0)',
              },
            },
          },
        },
      });

      render(
        <ThemeProvider theme={customTheme}>
          <LinearProgress data-testid="themed-linear-progress">
            Themed LinearProgress
          </LinearProgress>
        </ThemeProvider>
      );

      const element = screen.getByTestId('themed-linear-progress');
      expect(element).toBeInTheDocument();
    });

    it('supports dark mode', () => {
      const darkTheme = createTheme({
        palette: {
          mode: 'dark',
        },
      });

      render(
        <ThemeProvider theme={darkTheme}>
          <LinearProgress>Dark Mode LinearProgress</LinearProgress>
        </ThemeProvider>
      );

      expect(screen.getByRole('progressbar')).toBeInTheDocument();
    });
  });

  // Performance testing
  describe('Performance', () => {
    it('renders efficiently with many instances', () => {
      const manyComponents = Array.from({ length: 100 }, (_, i) => (
        <LinearProgress key={i}>LinearProgress {i}</LinearProgress>
      ));

      const { container } = renderWithTheme(<div>{manyComponents}</div>);
      expect(container.querySelectorAll('[role="progressbar"]')).toHaveLength(100);
    });

    it('respects reduced motion preferences', () => {
      // Mock prefers-reduced-motion
      Object.defineProperty(window, 'matchMedia', {
        writable: true,
        value: vi.fn().mockImplementation((query) => ({
          matches: query === '(prefers-reduced-motion: reduce)',
          media: query,
          onchange: null,
          addListener: vi.fn(),
          removeListener: vi.fn(),
          addEventListener: vi.fn(),
          removeEventListener: vi.fn(),
          dispatchEvent: vi.fn(),
        })),
      });

      renderWithTheme(
        <LinearProgress animate="false">
          No Animation LinearProgress
        </LinearProgress>
      );

      expect(screen.getByRole('progressbar')).toBeInTheDocument();
    });
  });

  // Edge cases and error handling
  describe('Edge Cases', () => {
    it('handles empty children gracefully', () => {
      renderWithTheme(<LinearProgress />);
      expect(screen.getByRole('progressbar')).toBeInTheDocument();
    });

    it('handles null children gracefully', () => {
      renderWithTheme(<LinearProgress>{null}</LinearProgress>);
      expect(screen.getByRole('progressbar')).toBeInTheDocument();
    });

    it('handles undefined props gracefully', () => {
      renderWithTheme(
        <LinearProgress 
          variant={undefined as any} 
          size={undefined as any}
        >
          Undefined Props
        </LinearProgress>
      );
      expect(screen.getByRole('progressbar')).toBeInTheDocument();
    });

  });

  // Snapshot testing
  describe('Snapshots', () => {
    it('matches snapshot for default props', () => {
      const { container } = renderWithTheme(
        <LinearProgress>Default LinearProgress</LinearProgress>
      );
      expect(container.firstChild).toMatchSnapshot();
    });

    it('matches snapshot for all variants', () => {
      const variants: LinearProgressProps['variant'][] = [
        'determinate',
        'indeterminate',
        'buffer',
        'query',
      ];

      variants.forEach((variant) => {
        const props: any = { variant };
        
        // Add required props for specific variants
        if (variant === 'determinate') {
          props.value = 50;
        } else if (variant === 'buffer') {
          props.value = 50;
          props.valueBuffer = 70;
        }
        
        const { container } = renderWithTheme(
          <LinearProgress {...props}>{variant} LinearProgress</LinearProgress>
        );
        expect(container.firstChild).toMatchSnapshot(`variant-${variant}`);
      });
    });
  });
});
