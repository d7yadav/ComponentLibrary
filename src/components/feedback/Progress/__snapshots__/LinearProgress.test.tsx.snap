// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`LinearProgress Component > Snapshots > matches snapshot for all variants > variant-buffer 1`] = `
<span
  aria-label="Progress"
  aria-valuemax="100"
  aria-valuemin="0"
  aria-valuenow="50"
  aria-valuetext="50%"
  class="MuiLinearProgress-root MuiLinearProgress-colorPrimary MuiLinearProgress-buffer css-1t0dxxh-MuiLinearProgress-root"
  role="progressbar"
>
  <span
    class="MuiLinearProgress-dashed MuiLinearProgress-dashedColorPrimary css-z7k20-MuiLinearProgress-dashed"
  />
  <span
    class="MuiLinearProgress-bar MuiLinearProgress-bar1 MuiLinearProgress-barColorPrimary MuiLinearProgress-bar1Buffer css-1igjkhl-MuiLinearProgress-bar1"
    style="transform: translateX(-50%);"
  />
  <span
    class="MuiLinearProgress-bar MuiLinearProgress-bar2 MuiLinearProgress-colorPrimary MuiLinearProgress-bar2Buffer css-1v1662g-MuiLinearProgress-bar2"
    style="transform: translateX(-30%);"
  />
</span>
`;

exports[`LinearProgress Component > Snapshots > matches snapshot for all variants > variant-determinate 1`] = `
<span
  aria-label="Progress"
  aria-valuemax="100"
  aria-valuemin="0"
  aria-valuenow="50"
  aria-valuetext="50%"
  class="MuiLinearProgress-root MuiLinearProgress-colorPrimary MuiLinearProgress-determinate css-18cps20-MuiLinearProgress-root"
  role="progressbar"
>
  <span
    class="MuiLinearProgress-bar MuiLinearProgress-bar1 MuiLinearProgress-barColorPrimary MuiLinearProgress-bar1Determinate css-l16vtb-MuiLinearProgress-bar1"
    style="transform: translateX(-50%);"
  />
</span>
`;

exports[`LinearProgress Component > Snapshots > matches snapshot for all variants > variant-indeterminate 1`] = `
<span
  aria-label="Progress"
  aria-valuemax="100"
  aria-valuemin="0"
  class="MuiLinearProgress-root MuiLinearProgress-colorPrimary MuiLinearProgress-indeterminate css-18cps20-MuiLinearProgress-root"
  role="progressbar"
>
  <span
    class="MuiLinearProgress-bar MuiLinearProgress-bar1 MuiLinearProgress-barColorPrimary MuiLinearProgress-bar1Indeterminate css-lcqqge-MuiLinearProgress-bar1"
  />
  <span
    class="MuiLinearProgress-bar MuiLinearProgress-bar2 MuiLinearProgress-barColorPrimary MuiLinearProgress-bar2Indeterminate css-lfb3c7-MuiLinearProgress-bar2"
  />
</span>
`;

exports[`LinearProgress Component > Snapshots > matches snapshot for all variants > variant-query 1`] = `
<span
  aria-label="Progress"
  aria-valuemax="100"
  aria-valuemin="0"
  class="MuiLinearProgress-root MuiLinearProgress-colorPrimary MuiLinearProgress-query css-12hj3fj-MuiLinearProgress-root"
  role="progressbar"
>
  <span
    class="MuiLinearProgress-bar MuiLinearProgress-bar1 MuiLinearProgress-barColorPrimary MuiLinearProgress-bar1Indeterminate css-lcqqge-MuiLinearProgress-bar1"
  />
  <span
    class="MuiLinearProgress-bar MuiLinearProgress-bar2 MuiLinearProgress-barColorPrimary MuiLinearProgress-bar2Indeterminate css-lfb3c7-MuiLinearProgress-bar2"
  />
</span>
`;

exports[`LinearProgress Component > Snapshots > matches snapshot for default props 1`] = `
<span
  aria-label="Progress"
  aria-valuemax="100"
  aria-valuemin="0"
  class="MuiLinearProgress-root MuiLinearProgress-colorPrimary MuiLinearProgress-indeterminate css-18cps20-MuiLinearProgress-root"
  role="progressbar"
>
  <span
    class="MuiLinearProgress-bar MuiLinearProgress-bar1 MuiLinearProgress-barColorPrimary MuiLinearProgress-bar1Indeterminate css-lcqqge-MuiLinearProgress-bar1"
  />
  <span
    class="MuiLinearProgress-bar MuiLinearProgress-bar2 MuiLinearProgress-barColorPrimary MuiLinearProgress-bar2Indeterminate css-lfb3c7-MuiLinearProgress-bar2"
  />
</span>
`;
